### PR TITLE
Make studio.code.org/s/express-2021/lessons/1/levels/2 work offline

### PR DESCRIPTION
**Pull request by daynew**
We want to setup studio.code.org/s/express-2021/lessons/1/levels/2 to able to work while the user is offline. This is currently hidden behind an experiment cookie in the user's browser.

## Links
* [JIRA](https://codedotorg.atlassian.net/browse/FND-1918) Initial implementation.
* [JIRA](https://codedotorg.atlassian.net/browse/FND-1908) Add retries and cleanup old data.

## Testing story
* Manually tested
  * Enable chrome://flags/#allow-insecure-localhost if you are testing on localhost.
  * Visit http://localhost-studio.code.org:3000/offline/join_pilot
  * Visit http://localhost-studio.code.org:3000/s/express-2021/lessons/1/levels/2
  * Open the developer tools and go into offline mode.
  * Refresh the page and observe that you can still complete the level.

### CODE REVIEW

File: apps/src/offline/offline-service-worker.js [hunk 1]
```diff
@@ -1,19 +1,54 @@
+import 'whatwg-fetch';
+
 /**
  * The Service Worker which enables the offline PWA experience for Code.org. It is responsible for
  * caching html, javascript, images, etc so that they can be used when students don't have an
  * internet connection.
  */
-const cacheVersion = 1;
+const cacheVersion = 2;
```
Comment (annaxuphoto) [inline apps/src/offline/offline-service-worker.js:10]: can we add documentation on when/why we increment cache version?

File: apps/src/offline/offline-service-worker.js [hunk 2]
```diff
@@ -1,19 +1,54 @@
...
-  e.waitUntil(
-    (async () => {
-      const cache = await caches.open(cacheName);
-      console.log('[Service Worker] Caching:', cachedFiles);
-      await cache.addAll(cachedFiles);
-    })()
+  // TODO - Add retry logic if cache.addAll fails for one of the assets.
+  e.waitUntil(caches.open(cacheName).then(cache => cacheFiles(cache)));
+});
+
```
Comment (annaxuphoto) [inline apps/src/offline/offline-service-worker.js]: when do we clear cache?

File: apps/src/offline/offline-service-worker.js [hunk 3]
```diff
@@ -1,19 +1,54 @@
...
+self.addEventListener('fetch', event => {
+  //Strip the query string parameters
+  const url = event.request.url.split('?')[0];
+  event.respondWith(
+    caches.open(cacheName).then(cache =>
+      cache.match(url).then(
+        response =>
+          // Return the cached response, otherwise fetch it over the internet.
+          response || fetch(event.request)
```
Comment (annaxuphoto) [inline apps/src/offline/offline-service-worker.js:41]: open question: do we do cache-first or network-first (or cache-only) approach?

Comment (annaxuphoto) [reply apps/src/offline/offline-service-worker.js:41]: also - constrain service worker actions to courses supported offline

**Review by annaxuphoto — COMMENTED** [review]
mostly documentation & questions. only blocking question is about constraining service worker actions to offline courses only (not all studio.code.org urls)

File: apps/src/offline/offline-service-worker.js (hunk 2 above)
Comment (daynew) [reply apps/src/offline/offline-service-worker.js]: Added a TODO

**Review by annaxuphoto — APPROVED** [review]
🎉

File: apps/src/offline/offline-service-worker.js [hunk 4]
```diff
@@ -5,27 +5,32 @@ import 'whatwg-fetch';
...
+const lessonName = 'express-2021-lessons-1';
+const cacheVersionName = `${pwaNamespace}-v${cacheVersion}`;
+const cacheName = `${cacheVersionName}-${lessonName}`;
 
+/**
+ * Runs when the service worker is updated or registered for the first time.
+ */
 self.addEventListener('install', e => {
   console.log('[Service Worker] installing offline service worker!');
   // TODO - Add retry logic if cache.addAll fails for one of the assets.
```
Comment (daynew) [inline apps/src/offline/offline-service-worker.js]:
```suggestion
```

File: apps/src/offline/offline-service-worker.js [hunk 5]
```diff
@@ -1,19 +1,118 @@
...
+const lessonName = 'express-2021-lessons-1';
+const cacheVersionName = `${pwaNamespace}-v${cacheVersion}`;
+const cacheName = `${cacheVersionName}-${lessonName}`;
+
+/**
+ * Runs when the service worker is updated or registered for the first time.
+ */
 self.addEventListener('install', e => {
   console.log('[Service Worker] installing offline service worker!');
+  // TODO - Add retry logic if cache.addAll fails for one of the assets.
```
Comment (daynew) [inline apps/src/offline/offline-service-worker.js]:
```suggestion
```

**Pull request merged**